zeshel/retrieval/model.py

```diff
@@ -0,0 +1,263 @@
+            self.cosine = nn.CosineSimilarity(dim=dim)
+        self.attn = attn
+        self.dim = dim
+        self.get_weights = get_weights
+        self.residual = residual
+
+    def forward(self, xs, ys, mask_ys=None, values=None):
+        """
+        Compute attention.
+        Attend over ys with query xs to obtain weights, then apply weights to
+        values (ys if yalues is None)
+        Args:
+            xs: B x query_len x dim (queries)
+            ys: B x key_len x dim (keys)
+            mask_ys: B x key_len (mask)
+            values: B x value_len x dim (values); if None, default to ys
+        """
+        if self.attn == 'cosine':
+            l1 = self.cosine(xs, ys).unsqueeze(self.dim - 1)
+        else:
+            l1 = torch.matmul(xs, ys.transpose(-1, -2))
+            if self.attn == 'sqrt':
+                d_k = ys.size(-1)
+                l1 = l1 / math.sqrt(d_k)
+        if mask_ys is not None:
+            attn_mask = (mask_ys == 0).unsqueeze(-2)
+            l1.masked_fill_(attn_mask, -NEAR_INF)
+        l2 = F.softmax(l1, -1, dtype=torch.float).type_as(l1)
+        if values is None:
+            values = ys
+        lhs_emb = torch.matmul(l2, values)
+
+        # # add back the query
+        if self.residual:
+            lhs_emb = lhs_emb.add(xs)
+
+        if self.get_weights:
+            return lhs_emb.squeeze(self.dim - 1), l2
+        else:
+            return lhs_emb
+
+
+class HardAttention(nn.Module):
+    """
+    Implements simple/classical hard attention.
+    """
+
+    def __init__(self):
+        super().__init__()
+
+    def forward(self, xs, ys):
+        """
+
+        :param xs: (B,T_x,d)
+        :param ys: (B,C,T_y,d)
+        :return: (B,C)
+        """
+        bsz, l_x, d = xs.size()
+        bsz, C, l_y, d = ys.size()
+        scores = (torch.matmul(xs, ys.reshape(bsz, -1, d).transpose(-1,
+                                                                    -2)).reshape(
+            bsz, l_x, C, l_y).max(-1)[0]).sum(1)
+        return scores
+
+
+class SoftAttention(nn.Module):
+    """
+    Implements simple/classical attention.
+    """
+
+    def __init__(self):
+        super().__init__()
+        self.attention = PolyAttention(dim=2, attn='basic',
+                                       get_weights=False)
+
+    def forward(self, xs, ys, values, mask_ys):
+        """
+
+        :param xs: (1,C,T_y,d)
+        :param ys: (B,T_x,d)
+        :param values: (B,T_x,d)
+        :param mask_ys: (B,T_x)
+        :return: (B,C)
+        """
+        bsz_x, C, l_y, d = xs.size()
+        xs = xs.reshape(bsz_x, -1, d)
+        bsz, l_x, d = ys.size()
+        attended_embeds = self.attention(xs, ys,
+                                         mask_ys=mask_ys,
+                                         values=values)  # (B,CT_y,d)
+        scores = (attended_embeds * xs).sum(-1).reshape(
+            bsz, C, l_y).sum(-1)
+        return scores
+
+
+class UnifiedModel(nn.Module):
```

Comment:
Just a suggestion for a more explicit name, maybe "UnifiedRetriever"? We have another model for reranking. In general it's good to make names as detailed as possible within clarity.